File: euler/haskell_solutions/nine.hs
-- Project Euler Problem Number Nine:
--
-- Find the product of the Pythagorean
-- triple elements whose sum is 1,000.

problemNine :: [(Integer, Integer, Integer)]
problemNine = [(a, b, c) | a <- [1..500], b <- [1..500], c <- [1..500], a^2 + b^2 == c^2, a + b + c == 1000]

-- problemNine ==> [(200,375,425),(375,200,425)]
--
-- The product is 31875000, but I'm more interested in the triple itself.
-- This program works, but is too slow—it takes several minutes to run.

File: euler/clojure_solutions/nine.clj
; Project Euler Problem Number Nine:
;
; Find the product of the Pythagorean
; triple elements whose sum is 1,000.

; Utility functions
(defn square [n]
  (* n n))

(defn triple? [a, b, c]
  (= (+ (square a) (square b)) (square c)))

(defn problemNine [n]
  (for [a (range 1 n)
        b (range a n)
        c [(- 1000 a b)]
        :when (triple? a, b, c)]
   (list, a, b, c)))

; (problemNine 1000) ==> ((200 375 425))
;
; The question asks for the product, which is 31875000,
; but I'm more interested in the triple itself.
;
